Adds a peer-departed state

lazypower opened the pull request:
Related to https://github.com/juju-solutions/layer-etcd/issues/5

Commit:
Fixes etcd scale-down issue

Related to https://github.com/juju-solutions/layer-etcd/issues/5

lazypower commented:
I'm on the fence about this. This code in the interface works as is, but @johnsca says the -broken relation in reactive is broken. I'm only using the state as it gets fired on the single machine departing the cluster, which triggers self destruction, i mean, de-registration.

mbruzek commented:
@johnsca What is the problem with the "-broken" hook that @chuckbutler talks about in the comment above?

mbruzek commented:
I would think you would want to do this in the "-departed" hook, giving the code time to unregister with the cluster. My understanding of "-broken" is they relationship is completely severed. I am not against using "-broken" here, but this looks like a perfect example of "-departed" to me.

The relevant code reactive code that uses this state:  https://github.com/juju-solutions/layer-etcd/blob/master/reactive/etcd.py#L315

mbruzek commented on peers.py:
```diff
@@ -34,6 +34,13 @@ def peers_changed(self):
 
+    @hook('{peers:etcd}-relation-broken')
+    def peers_going_away(self):
+        conv = self.conversation()
```
Couldn't you scope the conversation to UNIT right here and do this in '-departed'?

mbruzek commented:
@chuckbutler pointed me to the conversation about this bug/problem: https://lists.ubuntu.com/archives/juju/2015-December/006245.html

I am +1 to this code if it works (as I am sure it does), but I am concerned as my understanding of the "-departed" is appropriate here. We would want each unit to run some code to remove the dying unit from their "peer list" configuration, as well as the dying unit to "unregister" itself from the master/leader or cluster.

Commit:
State rework per @johnsca

Commit:
Adds dismiss method

Commit:
Adding the peer relation data setter and map return.

Commit:
Patched conversation behavior to reflect scope:unit when transmitting
cluster_id

johnsca commented:
LGTM

Commit:
Removing cluster_unit_id details in relation-data

The cluster_unit_id work was problematic as it was inconsistent in the leader getting the details. Instead this was reworked in https://github.com/juju-solutions/layer-etcd/pull/21 to leverage leader-state-membership data to find the guid and perform deregistration on the departing units behalf.

mbruzek commented on peers.py:
```diff
@@ -1,69 +1,42 @@
-                                public_address=public_address)
-        conversation.remove_state('{relation_name}.joining')
-        conversation.remove_state('{relation_name}.declare_self')
+        peers = []
```
There is a python one liner for these four lines. This way is fine, just flexing my python muscles.

mbruzek commented:
+1 to this change, I like the removal of not_unless and making it easy to understand what is going on.
